---
title: "otchet_part_1"
author: "anna kvach"
date: "12/16/2020"
output: 
  html_document:
    toc: true
    number_sections: true 
    toc_depth: 4  
    theme: united  
    highlight: tango  
---

# Obtain data

There are two options how to obtain data: with help of Sratoolkit and simle bash code or manually from EBLM-EBI. 

## Sratoolkit + simple bash 

Using this method, we created a simple bash program that receives a file.txt with sra identifiers as input. Our code uses the prefetch and fastq-dump programs from the Sratoolkit package, and as a result downloads forward and backward reads.


First of all we created a file.txt with all sra id on each line.

```{bash, content of txt.file, eval=F, echo=T}
vi id.txt

SRR11096622
SRR11096623
SRR11096624
SRR11096625
SRR11096626
SRR11096627
SRR11096628
SRR11096629
SRR11096630
SRR11096631
SRR11096632
SRR11096633
SRR11096634
SRR11096635
SRR11096636
SRR11096637
SRR11096638
SRR11096639
SRR11096640
```



After that sra_dowload_with_sratoolkit.sh was create.  

```{bash, sra_dowload_with_sratoolkit, eval=F, echo=T}
#!/bin/bash  
# for SRA files downloading with help of sratoolkit                                           
FILE=$1                            
while read id                      
do                                 
prefetch $id                       
echo "for $id prefetch done"       
fastq-dump --split-files --split-3 ${id}.sra
echo "for $id fastq-dump done"     
done < $FILE
```

And after that:

```{bash, eval=F, echo=T}
sra_dowload_with_sratoolkit.sh sra_id.txt
```

In result you got for all samples SRRXXXXXXXX_1.fastq.gz (forward reads) and (revers reads) SRRXXXXXXXX_2.fastq.gz, and for some samples - SRRXXXXXXXX.fastq.gz - files without prefix. Files with no prefix are files of the remaining unpaired reads, and we did not use them.

```{bash, eval=F, echo=T}
mkdir forward_and_revers_raw_reads # create dir for forward and revers raw reads
mkdir bad_single_reads # create dir for unpaired reads

mv *_*.*  forward_and_revers_raw_reads # dir for forward and revers raw reads - data for future analisis
mv *.f* bad_single_reads # dir for remaining unpaired reads - did not use them
```


## Directly and manually from EBLM-EBI

Also data could be obtained directly and manually from EBLM-EBI - <https://www.ebi.ac.uk/ena/browser/view/PRJNA607082>. 

You should dowload forward and revers reads. Files with forward reads have "_1" prefix (for ex. SRR11096622_1.fastq.gz) and files with reverse reads have "_2" prefix (for ex. SRR11096622_2.fastq.gz). Files with no prefix are files of the remaining unpaired reads, and we did not need them.

# First data quality assessment FastQC and MultiQC

## FastQC

```{bash, eval=F, echo=T}
cd forward_and_revers_raw_reads # start to work with good paired reads

gunzip *fastq.gz # to unarchive all data  
mkdir FastQC_1_for_raw_data # create new dir for fastqc reports
fastqc -o * # create fastqc reports
```

## MultiQC

Quality reports show problems with:

* 

* 

# Trim Galore + FastQC after it

As we faced with adapters-problem we create file.fasta with adapters from <http://docs.blast2go.com/user-manual/tools-(pro-feature)/fastq-quality-check/#FASTQQualityCheck-PerBaseSequenceQuality>.

```{bash, eval=F, echo=T}
vi adapters.fasta

>Illumina_Universal_Adapter 
AGATCGGAAGAG
>Illumina_Small_RNA_3_Adapter
TGGAATTCTCGG
>Illumina_Small_RNA_5_Adapter
GATCGTCGGACT
```


```{bash, eval=F, echo=T}
#####
#!/bin/bash  
# for trimming with TrimGalore for forward and reverse reads                                           
FILE=$1                            
while read id                      
do                                     
/path_to/TrimGalore-0.6.6/trim_galore --paired --phred33 -q 20 --length 36 -stringency 1 -e 0.1 -a2 "file:/path_to/adapters.fasta" -o after_trim_galore_1  --fastqc {$1}_1.fastq {$id}_2.fastq 
echo "for $id fastq-dump done --split-3"     
done < $FILE
```

* -q 20 - remove low-quality reads with a Phred score cutoff of 20

* --length 36 - a minimum read length threshold of 36 bp

* -stringency 1 - a stringency parameter of 1 for adapter sequence overlap

* -e 0.1 - a maximum allowed error rate of (0.1) 

* -a2 "file:/path_to/adapters.fasta" - to specify individual adapter sequences from our file.fasta for the two reads of paired-end files

* -o after_trim_galore_1  - all output will be written to this directory 

* --fastqc - create fastQC report

* {$1}_1.fastq - forward reads 

* {$id}_2.fastq - reverse reads 















# References

FastQC - 

MultiQC - <https://multiqc.info/>

Sratoolkit - <http://ncbi.github.io/sra-tools/ >

TrimGalore-0.6.6 - 

